File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        int low =0;
        int high =n-1;
        int mid;
        while(low <= high){
            int mid = (low+high)/2;
            // If target is present at the middle index
            if(nums[mid] == target){
                cout<<mid<<endl;    
                return mid;
            }

            // Check which half is sorted
            //Left Sorted
            if(nums[low] <= nums[mid]){
                if(nums[low] <= target && target <= nums[mid]){
                    high = mid-1;
                    
                }
                else{
                    low = mid+1;
                }
            }
            // Right Sorted
            else{
                if(nums[mid] <= target &&target <= nums[high]){
                    low = mid+1;
                    cout<<mid<<" "<<low<<endl;
                }
                else{
                    high = mid-1;
                    cout<<mid<<" "<<high<<endl;
                }
            }
        }
        cout<<mid<<endl;
        return -1;
    }
};

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.java
class Solution {
    public int binarySearch(int[] nums, int target, int left, int right){
        int n = nums.length;

        while(left <= right){
            int mid = left + (right - left)/2;

            if(nums[mid] == target){
                return mid;
            }

            else if(nums[mid] > target){
                right = mid - 1;
            }

            else{
                left = mid + 1;
            }
        }

        return -1;
    }

    public int search(int[] nums, int target) {
        int n = nums.length;

        int left = 0;
        int right = n-1;

        while(left < right){
            int mid = left + (right - left)/2;

            if(nums[mid] > nums[right]){
                left = mid + 1;
            }else{
                right = mid;
            }

            if(left == right){
                // We get the point of rotation
                break;
            }
        }

        left = 0;
        int ans = -1;

        ans = binarySearch(nums, target, left, right-1);
        if(ans != -1) return ans;

        ans = binarySearch(nums, target, right, n-1);
        return ans;
    }
}